docs: initial content for Updating a NativeScript app (#17)
NativeScript/docs: content/guide/updating-an-app.md

---
title: Updating an app
---

> To upgrade a NativeScript application you need to upgrade several things: NativeScript CLI Tooling, the iOS and Android runtimes and the `@nativescript/core` module. In the steps below you will see how to do this.

Reviewer: ```suggestion
To upgrade a NativeScript app, ensure to update the NativeScript CLI Tooling, iOS and Android runtimes, and @nativescript/core. The following steps demonstrate the upgrade process.
```

To get the latest version of the NativeScript CLI, run:

```cli
npm install -g nativescript
```

To upgrade an app, run:

```cli
ns update
```

In order to get the latest development release instead, pass **next** as argument:

```cli
ns update next
```

> You can also switch to specific version by passing it to the command:

Reviewer: ```suggestion
To switch to a specific version by passing it to the command:
```

```cli
ns update 8.0.0
```

::: tip Note
The command `ns update` is updating the `@nativescript/core`, `@nativescript/webpack`, and the runtimes (`@nativescript/android`and`@nativescript/ios`). The command is combining the next three commands in this article (`ns platform add`, `npm i --save @nativescript/core`and`npm i @nativescript/webpack --save-dev`).

:::

::: warning Important
When using the `--configs` flag, any previous configuration will be overwritten and lost. Consider saving any custom code that you have introduced in your `webpack.config.js` and reapplying the code after using the `--configs` flag.
:::

#### Upgrading platforms

> Follow those steps in order to get the latest versions of Android and/or iOS runtimes. Navigate to the root folder of your project, and then if you are working on a Android project, type:

Reviewer: ```suggestion
Follow those steps in order to get the latest versions of Android and/or iOS runtimes.  Navigate to the root folder of your project, and then if working on a Android project, type:
```

```cli
ns platform remove android
ns platform add android
```

> and/or (if you are working on a iOS version on a Mac):

Reviewer: ```suggestion
If working on a iOS version using a Mac OS:
```

```cli
ns platform remove ios
ns platform add ios
```

#### Upgrading @nativescript/core

The cross-platform modules are available as a npm package named [@nativescript/core](https://www.npmjs.com/package/@nativescript/core).

> In order to use them in your project, you will have to explicitly install the package, for example (assuming you are still in your main app project folder from the steps above):

Reviewer: ```suggestion
To utilize these packages in your project, they must be installed explicitly.  For instance, assuming you are in your main app project folder from the previous steps, follow the instructions below.
```

```cli
npm install @nativescript/core@latest --save
```

This installs the **@nativescript/core** package to the node_modules folder and adds it as a dependency to the package.json of the project.

::: warning Important

> The `ns create` command will create a new project, add the **@nativescript/core** package as a dependency to its package.json and install it. So each new project you create will have the **@nativescript/core** package installed and you do not have to install it explicitly.

Reviewer: ```suggestion
The ns create command creates a new project, adds the @nativescript/core package as a dependency to its package.json, and automatically installs it.  Thus, each new project created will have the @nativescript/core package installed without requiring explicit installation.
```

:::

Another place to find **@nativescript/core** package is [NativeScript Releases](https://github.com/NativeScript/NativeScript/releases/), where you can find a collection of the available @nativescript/core-\*.tgz packages for every release. You can download a selected release and install it by running: `npm install <path to @nativescript/core-*.tgz> --save`.

#### Upgrading Angular dependencies

> The Angular plugin is available as an npm package named [@nativescript/angular](https://www.npmjs.com/package/@nativescript/angular). To update the version of the plugin and the related dependency, the package should be explicitly installed, and the related Angular dependencies should be updated accordingly. To ease the update process, the plugin comes with an automated script `update-app-ng-deps` located in `<project-folder/node_modules/.bin>` folder.

Reviewer: ```suggestion
The Angular plugin is available as [@nativescript/angular](https://www.npmjs.com/package/@nativescript/angular). To update the version of the plugin and the related dependency, the package should be explicitly installed, and the related Angular dependencies should be updated accordingly. To ease the update process, the plugin comes with an automated script `update-app-ng-deps` located in `<project-folder/node_modules/.bin>` folder.
```

```cli
npm i @nativescript/angular@latest --save
./node_modules/.bin/update-app-ng-deps
npm i
```

---

title: Running Latest Code
description: NativeScript Documentation - Running Latest Code
position: 40
slug: latest-code
previous_url: /running-latest

---

#### Running the Latest Code

> Often when working with open-source projects, one needs functionality that has not yet passed the full release cycle, or even functionality that is not yet fully implemented. We know that many of you are experimenters and want to try the latest and greatest features of NativeScript. That is why we tried to make this process simple and easy to follow. There are two ways to get the latest development code for NativeScript:

Reviewer: ```suggestion
Often when working with open-source projects, at times there is a requirement for functionality that has not yet passed the full release cycle, or even functionality that is not yet fully implemented. We know that many developers using open-source software are experimenters and would wish to enjoy the option to try out the latest and greatest features of NativeScript. This is why we tried to make this process simple and easy to follow. There are two ways to get the latest development code for NativeScript:
```

- You can get it via npm.
- You can build the source code.

#### Getting the latest development version via npm

> As an open-source project NativeScript keeps not only its source code but its build infrastructure open. That is why we choose [Travis CI](https://travis-ci.org/) for our nightly builds. Every commit in the master branch of all major NativeScript repos triggers a [Travis CI](https://travis-ci.org/) build which publishes an npm package that can be used directly. Follow those simple steps to get the latest development version of NativeScript:

Reviewer: ```suggestion
As an open-source project NativeScript keeps not only its source code, but its build infrastructure open. Every commit in the main branch of all major NativeScript repos triggers a Github Action Workflow that publishes a npm package, which can be used directly.  Follow these steps in order to get the latest development version of NativeScript:
```

- Uninstall any existing NativeScript versions:

```cli
npm uninstall -g nativescript
```

- Install the latest development version of NativeScript CLI:

```cli
npm install -g nativescript@next
```

- Edit the package.json file in your project and replace @nativescript/core, @nativescript/android and @nativescript/ios versions with `next`:

```json
{
  "description": "NativeScript Application",
  "dependencies": {
    "@nativescript/core": "next"
  },
  "devDependencies": {
    "@nativescript/android": "next",
    "@nativescript/ios": "next"
  }
}
```

> Instead of editing the package.json file by hand, you could run the following commands:

Reviewer: ```suggestion
Instead of editing the package.json file directly, the following commands can be run:
```

```cli
ns platform add ios@next
ns platform add android@next
ns plugin add @nativescript/core@next
```

- Run the `npm install` command to update the node modules:

```cli
cd <your-project-folder>
npm install
```

You are now ready to use the latest development version of NativeScript.

#### Building the source code

##### Reasoning

> <!-- TODO: fix links -->

Reviewer: Link below on 155 is good

Having said that, each subfolder of the [apps](https://github.com/NativeScript/NativeScript/tree/master/apps) subfolder of the repo represents a single application.

#### Using the latest

To use the latest:

- Build the repo.
- Navigate to your project folder.
- Delete the `@nativescript/core` folder from the `node_modules` subfolder of your project (i.e., `rm -rf node_modules/@nativescript/core` for Linux or `rd /S /Q node_modules\@nativescript/core`).
- Install the newly built package (`npm install [PATH-TO-NATIVESCRIPT-REPO/bin/dist/nativescript-core-x.x.x.tgz]`).

#### Handling internal breaking changes

> It is possible that an internal breaking change gets introduced involving an update to both the runtimes and the modules. An internal breaking change would mean that the public API of the tns_modules does not get affected, but a work in progress change in the runtimes requires a change in the internal code of the tns_modules themselves.

Reviewer: ```suggestion
It is possible for an internal breaking change to be introduced, affecting both the runtimes and core.  This type of change requires modifications to the internal code of @nativescript/core, while the public API remains unaffected.
```

When such a case happens, the [ios](https://github.com/NativeScript/ns-v8ios-runtime) and [android](https://github.com/NativeScript/android-runtime) runtimes must be built separately and updated via the CLI command of:
`ns platform update android/ios --frameworkPath=[Path-to-Runtime-Package]`

#### Building the runtimes

> As the NativeScript framework gets distributed via npm, the runtimes are also packed as npm packages. For consistency reasons, the native builds (gradle/xcode-build) are wrapped by grunt builds that do the job.

Reviewer: ```suggestion
The NativeScript framework and runtimes are distributed as npm packages. To maintain consistency, the native builds (gradle/xcode-build) are encapsulated within grunt builds.
```

#### Building the Android runtime

The [android runtime](https://github.com/NativeScript/android-runtime) depends on the [android-metadata-generator](https://github.com/NativeScript/android-metadata-generator).

> Provided you have all the dependencies set, the easiest way to have the Android runtime built is to clone the two repos to a single folder so that the two are sibling folders, `cd` into the `android-runtime` folder and run:

Reviewer: ```suggestion
If all dependencies are set, the simplest way to build the Android runtime is by cloning the two repositories into a single folder, ensuring they are sibling folders. Then navigate to the `android-runtime` directory and execute the specified command.
```

```cli
gradle packar -PwidgetsPath=./widgets.jar
```

The resulting @nativescript/android-x.x.x.tgz package will get created in the `dist` folder.

#### Building the iOS runtime

Follow the instructions on setting up the dependencies for building the [ios runtime](https://github.com/NativeScript/ns-v8ios-runtime) in the repository README and then run `grunt package`.

> The build @nativescript/ios-x.x.x.tgx package will get created in the `dist` folder.

Reviewer: ```suggestion
The @nativescript/ios-x.x.x.tgx package will be generated in the dist folder.
```
